### PR TITLE
update ERC721 template

### LINKED ISSUE #654: Validate metadata publishing on chain
Update tests, because ERC721 setmetadata has changed

### LINKED ISSUE #655: Validate metadata endpoint
When validating a DDO, aquarius should:
- respond with a 400 Header if DDO is not valid, and have errors in the body (json reponse)
- respond with a 200 Header if DDO is valid, adding the following to the output (json)
     - hash (sha256) of post data(ddo as octet)  (taken as bytes from input, no json.loads/dumps conversion)
     - signature (v, r, s)        - aquarius signature on hash, using aqua's private key
     - public_key

### PR DESCRIPTION
**Pull request by alexcos20**
Closes #654 
Closes #655

### CODE REVIEW

**Comment by alexcos20** [comment]
so, we hit a circular dependency.  
Aqua needs Provider (with updated code).  But in order to update Provider, we need Aqua with updated code.

So both PRs ([this ](https://github.com/oceanprotocol/aquarius/pull/656) and https://github.com/oceanprotocol/provider/pull/311) should be merged in the same time.

To test it , do the following:
 - clone provider, build your own image
 - clone aquarius, build your own image
 - update barge to use those images , instead of default aqua & provider
 - Run the tests
 - Approve the PR

**Review by DMats — COMMENTED** [review]
I followed the instructions https://github.com/oceanprotocol/aquarius/pull/656#issuecomment-1001662168 and also tweaked the IP addresses so the tests could connect to barge, but was unable to make the Aquarius tests pass.

To make the Aquarius tests connect to barge, I had to update `pytest.ini` like so:

```patch
-    D:DB_HOSTNAME = 127.0.0.1
+    D:DB_HOSTNAME = 172.15.0.6
    ...
-    D:EVENTS_RPC = http://127.0.0.1:8545
-    D:NETWORK_URL = http://127.0.0.1:8545
+    D:EVENTS_RPC = http://172.15.0.3:8545
+    D:NETWORK_URL = http://172.15.0.3:8545
```

EDIT: 

The errors look similar to the ones that appear in the Github checks. (with the exception of `test_get_web3_connection_provider` which fails because of the tweak described above)

```pytest
========================================================================== short test summary info ===========================================================================
FAILED tests/test_assets_api.py::test_resolveByDtAddress - assert 0 > 0
FAILED tests/test_assets_api.py::test_get_assets_names - AssertionError: did name not found.
FAILED tests/test_events.py::test_publish_and_update_ddo - KeyError: 'id'
FAILED tests/test_events.py::test_publish_and_update_ddo_with_lzma - KeyError: 'id'
FAILED tests/test_events.py::test_publish_and_update_ddo_with_lzma_and_encrypt - KeyError: 'id'
FAILED tests/test_events.py::test_publish_and_update_ddo_unencrypted - KeyError: 'id'
FAILED tests/test_events.py::test_publish - KeyError: 'id'
FAILED tests/test_events.py::test_publish_and_update_ddo_rbac - KeyError: 'id'
FAILED tests/test_events.py::test_get_last_processed_block - assert 31 == 0
FAILED tests/test_events.py::test_order_started - KeyError: 'stats'
FAILED tests/test_processors.py::test_process_fallback - KeyError: 'id'
FAILED tests/test_purgatory.py::test_purgatory_with_assets - KeyError: 'stats'
FAILED tests/test_purgatory.py::test_purgatory_with_accounts - KeyError: 'stats'
FAILED tests/test_util.py::test_get_web3_connection_provider - AssertionError: assert 'http://172.15.0.3:8545' == 'http://127.0.0.1:8545'
```

File: aquarius/app/assets.py [hunk 1]
```diff
@@ -4,17 +4,24 @@
 #
 import json
 import logging
+import os
+from hashlib import sha256
```
Comment (calina-c) [inline aquarius/app/assets.py:8]: please alphabetize imports in groups :)

File: aquarius/app/assets.py [hunk 2]
```diff
@@ -305,15 +312,38 @@ def validate_remote():
...
+                ]
+            ),
+            400,
+        )
+
+    pk = os.environ.get("PRIVATE_KEY", None)
+    values = {"hash": "", "publicKey": "", "r": "", "s": "", "v": ""}
+    if pk:
+        raw = request.get_data()
+        wallet = Account.from_key(private_key=pk)
```
Comment (calina-c) [inline aquarius/app/assets.py:336]: can you please name this provider_wallet like everywhere else? I also think there is an util callled get_provider_wallet or provider_wallet that would help here

File: aquarius/app/assets.py [hunk 3]
```diff
@@ -305,15 +312,38 @@ def validate_remote():
...
+                    }
+                ]
+            ),
+            400,
+        )
+
+    pk = os.environ.get("PRIVATE_KEY", None)
+    values = {"hash": "", "publicKey": "", "r": "", "s": "", "v": ""}
+    if pk:
+        raw = request.get_data()
```
Comment (calina-c) [inline aquarius/app/assets.py:335]: my opinion is to move the contents of this if to a different function and only call it later, when we need to return it (line 362 currently, when doing "if valid")

File: tests/helpers.py [hunk 4]
```diff
@@ -162,3 +162,14 @@ def run_request(client_method, url, data=None):
         )
 
     return _response
+
+
+def run_request_octet(client_method, url, data=None):
+    if data is None:
+        _response = client_method(url, content_type="application/octet-stream")
```
Comment (calina-c) [inline tests/helpers.py:169]: I think you can return this directly, without assigning `_response`. Then later, you can remove the "else" keyword and return the other value directly.

File: tests/test_validate_api.py [hunk 5]
```diff
@@ -2,7 +2,8 @@
 # Copyright 2021 Ocean Protocol Foundation
 # SPDX-License-Identifier: Apache-2.0
 #
-from tests.helpers import run_request
+import json
+from tests.helpers import run_request_octet
```
Comment (calina-c) [inline tests/test_validate_api.py:6]: here as well, please organise imports.

**Review by calina-c — CHANGES_REQUESTED** [review]
just a few of my minor quirks, please :)
otherwise looks great!

**Review by calina-c — APPROVED** [review]

**Pull request merged**